### ISSUE TITLE
clang14 crashes compiling xscreensaver for riscv64

### ISSUE THREAD

**Issue by kraj** [llvm/llvm-project#53913]
attached testcase works ok with clang13 but clang14 we see a segfault.

version info
```
clang version 14.0.0 (https://github.com/llvm/llvm-project b2ca48a8412270e323ac8441772d0d69190a8417)
Target: riscv64-yoe-linux
```

stackdump
```
1.	<eof> parser at end of file
2.	Code generation
3.	Running pass 'Function Pass Manager' on module 'ant.c'.
4.	Running pass 'RISCV DAG->DAG Pattern Instruction Selection' on function '@draw_anant'
```

[ant-1d7471.tar.gz](https://github.com/llvm/llvm-project/files/8091906/ant-1d7471.tar.gz)

**Comment by DimitryAndric** [comment]
As far as I can see, it gets into some endless loop, allocating more and more memory, until it gets OOM-killed. This is usually tricky to reduce. :)

**Comment by asb** [comment]
There might be multiple issues at play here, but the given test case seems to be impacted by #53831. Compilation with `./bin/clang -DSTANDALONE -DHAVE_CONFIG_H -target riscv64 -march=rv64imafdc -O2 ~/llvm-project/ant-1d7471.c` hangs using current HEAD llvm/clang, but when disabling the problematic transform, it gets to the linking stage. I'm not seeing steadily increasing memory usage.

**Comment by asb** [comment]
This is fixed for me by c5bcfb983e47167a8a1826c1a64d7aa1849add06 (which matches my experience of reducing this test case and finding the same root cause). Thanks for reporting this, I'll close it as a dupe in favour of #53831.
